### ISSUE TITLE
Inconsistence between per-site result and per-read result

### ISSUE THREAD

**Issue by PanZiwei** [nanoporetech/modkit#46]
I have different results for per-site results using 2 different methods: 
a) Use the `modkit pileup` to extract per-site results directly 
b) Use the `modkit extract` to extract per-read result -> Acculate the per-read to per-site by reference genome location
Ideally, they should give the same results as the logic is the same.

So I have the following questions: 
1. What is the logic/relationship between your design for `modkit pileup` and `modkit extract`? Have you ever compared the results with the per-read and per-site?

2. Does base calling quality matter for the methylation calling? Do you apply any filter on the  `base_qual` for the `modkit pileup` per-site result? I know `mod_qual` is considered to calculate the base modification threshold. Not sure if a similar process is done for the base calling threshold. 

Thank you so much for your help!

**Comment by marcus1487** [comment]
The same logic applied to the `modkit extract` output should indeed produce the same results as the `modkit pileup` command. `base_qual` is not taken into account in pileup. 

Could you provide an example of a site where these values disagree? Or could you post the code you've used to apply the `mod_qual` threshold and perform the pileup logic?

**Comment by PanZiwei** [comment]
Thanks for the quick response. I believe the base quality filter in my per-read analysis caused the inconsistency. I will double-check and let you know once I have more ideas about it.  

BTW, is there a specific reason that `base_qual` is not considered in the process? Does base calling quality affect the methylation calling? It seems that for [latest R10.4.1 version it can achieve Q20](https://nanoporetech.com/accuracy), so I am not sure whether base_qual should be considered in this process.

Thank you so much for your help!

**Comment by marcus1487** [comment]
The correlation of base calling quality averaged over a whole read to mapping accuracy is quite good, but at the single base resolution the correlation is much weaker as one would expect. Additionally modified bases are more likely to decrease the canonical base calling score and would thus be very likely to bias the pileup results which would not be beneficial for most all analyses. For these reasons we do not use base quality in modified base pileup.

**Comment by zukailiu** [comment]
Update: I am using the same input_bam and 0.8/0.2 as the mod_threshold cutoff but still get slightly different results with the 2 methods. I didn't put any base quality filter. Any ide about the inconsistency? Thanks!

Method 1: Use the modkit pileup to extract per-site results directly
`modkit pileup -t ${THREAD} --cpg --ref ${ref_fasta} --filter-threshold 0.8 --mod-threshold m:0.8 ${input_bam} ${output_bed1}`

Method 2:

a) Use the modkit extract to extract per-read result
```code
medkit motif-bed ${ref_fasta} CG 0 > ${cg_bed}
modkit extract ${input_bam} ${output_tsv} --mapped-only --force --ref ${ref_fasta} --include-bed ${cg_bed} 
```

b) Use 0.8/0.2 -> Acculate the per-read to per-site by reference genome location
```
df=pd.read_csv(output_tsv, header=True, index =None
df_selected = df[(df['mod_qual']>= 0.8) | (df['mod_qual']<= 0.2)]
# Create a new column 'mod_qual' based on 'prob_mod' column
df_selected['mod_label'] = df_selected['mod_qual'].apply(lambda x: 1 if x > 0.8 else 0)

# Group by the specified columns and aggregate the data
df_grouped = df_selected.groupby(['chrom', 'start', 'end', 'ref_strand']).agg({
    'read_id': 'count',           # Count the number of reads for each group (Nvalid)
    'Ncanon': 'sum',         # Sum of [mod_label]=0 for each group
    'Nmod': 'sum',         # Sum of [mod_label]=1 for each group
}).reset_index()

# Calculate the methylation frequency
df_grouped['Nvalid'] = df_grouped['Ncanon'] + df_grouped['Nmod']
df_grouped['meth_freq'] = df_grouped['Nmod'] / df_grouped['Nvalid']
df_grouped.to_csv(output_bed2, header=True, index=None, sep='\t')
```

Then I compare the `${output_bed1}` and `${output_bed2}`
```
(base) [c-panz@sumner-log1 guppy6.3.8_site]$ head -5 ${output_bed1}
custom_DNA	130	131	m	18.18	+	77	14	63
custom_DNA	131	132	m	9.21	-	76	7	69
custom_DNA	191	192	m	10.0	+	80	8	72
custom_DNA 192	193	m	13.04	-	92	12	80
custom_DNA 238	239	m	11.84	+	76	9	67

(base) [c-panz@sumner-log1 guppy6.3.8_site]$ head -5 ${output_bed2}
custom_DNA	130	131	m	18.05	+	72	13	59
custom_DNA	131	132	m	9.21	-	76	7	69
custom_DNA	191	192	m	10.0	+	80	8	72
custom_DNA	192	193	m	13.19	-	91	12	79
custom_DNA	238	239	m	12.0	+	75	9	66
```

**Comment by PanZiwei** [comment]
Update: I am using the same input_bam and 0.8/0.2 as the mod_threshold cutoff but still get slightly different results with the 2 methods. I didn't put any base quality filter. Any ide about the inconsistency? Thanks!

Method 1: Use the modkit pileup to extract per-site results directly
```
modkit pileup -t ${THREAD} --cpg --ref ${ref_fasta} --filter-threshold 0.8 --mod-threshold m:0.8 ${input_bam} ${output_bed1}
```

Method 2:

a) Use the modkit extract to extract per-read result
```
medkit motif-bed ${ref_fasta} CG 0 > ${cg_bed}
modkit extract ${input_bam} ${output_tsv} --mapped-only --force --ref ${ref_fasta} --include-bed ${cg_bed} 
```
b) Use 0.8/0.2 -> Acculate the per-read to per-site by reference genome location
```
df=pd.read_csv(output_tsv, header=True, index =None
df_selected = df[(df['mod_qual']>= 0.8) | (df['mod_qual']<= 0.2)]
# Create a new column 'mod_qual' based on 'prob_mod' column
df_selected['mod_label'] = df_selected['mod_qual'].apply(lambda x: 1 if x > 0.8 else 0)

# Group by the specified columns and aggregate the data
df_grouped = df_selected.groupby(['chrom', 'start', 'end', 'ref_strand']).agg({
    'read_id': 'count',           # Count the number of reads for each group (Nvalid)
    'Ncanon': 'sum',         # Sum of [mod_label]=0 for each group
    'Nmod': 'sum',         # Sum of [mod_label]=1 for each group
}).reset_index()

# Calculate the methylation frequency
df_grouped['Nvalid'] = df_grouped['Ncanon'] + df_grouped['Nmod']
df_grouped['meth_freq'] = df_grouped['Nmod'] / df_grouped['Nvalid']
df_grouped.to_csv(output_bed2, header=True, index=None, sep='\t')
Then I compare the ${output_bed1} and ${output_bed2}
```

(base) [c-panz@sumner-log1 guppy6.3.8_site]$ head -5 ${output_bed1}
custom_DNA	130	131	m	18.18	+	77	14	63
custom_DNA	131	132	m	9.21	-	76	7	69
custom_DNA	191	192	m	10.0	+	80	8	72
custom_DNA 192	193	m	13.04	-	92	12	80
custom_DNA 238	239	m	11.84	+	76	9	67

(base) [c-panz@sumner-log1 guppy6.3.8_site]$ head -5 ${output_bed2}
custom_DNA	130	131	m	18.05	+	72	13	59
custom_DNA	131	132	m	9.21	-	76	7	69
custom_DNA	191	192	m	10.0	+	80	8	72
custom_DNA	192	193	m	13.19	-	91	12	79
custom_DNA	238	239	m	12.0	+	75	9	66

**Comment by ArtRand** [comment]
Hello @PanZiwei,

In the example data you've provided, the three rows that are different:
```
custom_DNA	130	131	m	18.18	+	77	14	63
custom_DNA      192	193	m	13.04	-	92	12	80
custom_DNA      238	239	m	11.84	+	76	9	67

// vs

custom_DNA	130	131	m	18.05	+	72	13	59
custom_DNA	192	193	m	13.19	-	91	12	79
custom_DNA	238	239	m	12.0	+	75	9	66
```
I suspect the difference is somewhere in how the python is filtering out base modification calls and how `modkit pileup` does. 

Would you be able to share the BAM from which you're seeing the difference? It can be just reads aligning to the positions you've shown above `samtools view -bh -L <bed_file> <bam> > <subset_bam>` will do it. I'd be happy to take a closer look.


The python code you've provided is incomplete, so I tried to fill in the gaps below

```python
X = pd.read_csv("test_reads_ignore_h_extract.tsv", delim_whitespace=True)
X_filt = X[(X["mod_qual"] >= 0.8) | (X["mod_qual"] <= 0.2)]

# Create a new column 'mod_qual' based on 'prob_mod' column
X_filt["mod_label"] = X_filt["mod_qual"].apply(lambda x: True if x > 0.8 else False)


# Group by the specified columns and aggregate the data
X_agg = X_filt.groupby(["chrom", "ref_position", "ref_strand"]).apply(
    lambda x: pd.Series({
        "strand": x["ref_strand"].unique()[0],
        "n_valid": x.shape[0],
        "n_mod": x["mod_label"].sum(),
        "n_canon": (~x["mod_label"]).sum(),

    })
).reset_index().rename(columns=dict(ref_position="start"))
X_agg["end"] = X_agg["start"] + 1
X_agg["pct_mod"] = X_agg["n_mod"] / X_agg["n_valid"]
[test_extract.zip](https://github.com/nanoporetech/modkit/files/12553782/test_extract.zip)

X_agg = X_agg[["chrom", "start", "end", "strand", "n_valid", "pct_mod", "n_mod", "n_canon"]]
```

It may not be perfect. Maybe you can try it and see if you get different answers. I made the file `test_reads_ignore_h_extract.tsv` by running `modkit adjust-mods --ignore h` on the [test reads](https://github.com/nanoporetech/modkit/raw/master/tests/resources/bc_anchored_10_reads.sorted.bam). Followed by `modkit extract`, I've also attached the file here. [test_extract.zip](https://github.com/nanoporetech/modkit/files/12553777/test_extract.zip)

**Comment by PanZiwei** [comment]
Hi thanks for the instruction. I followed your code but the inconsistency is still the same. I checked the per-read result after `modkit extract` and per-site result after `modkit pileup`, and I believe that **The `--mod-threshold` didn't filter the result**

More specifically, 73 reads covered the 1162084 location, and only 68 reads passed the 0.8/0.2 threshold at read level. However,  the `modkit pileup` calculates all reads.

I am using modkit `v0.1.12` for all the running. I have attached a test folder including the bam file and all scripts for your reference.
[test.zip](https://github.com/nanoporetech/modkit/files/12554886/test.zip)

```
## modkit pileup
custom_dna	1162084	1162085 16.44	+	73	12	61
custom_dna	1162085	1162086	9.21	 -	76	7	69
custom_dna	1162145	1162146	9.46	+	74	67	7
custom_dna	1162145	1162146	13.19	+	91	12	79
custom_dna	1162145	1162146	10.00	+	70	7	63

// vs
## modkit extract
custom_dna	1162084	1162085 0.161765	+	68	11 57
custom_dna	1162085	1162086	0.092105	 -	76	7  69	
custom_dna	1162145	1162146	0.094595	+	74	7   67
custom_dna   1162146   1162147   0.133333 -  90  12 78
custom_dna  1162192 1162193  0.101449 +  69  7  62
```

Per-read result after `modkit extract`:
```
input_path='/fastscratch/c-panz/test'
input_bed='subset.1162084.read.tsv'
df = pd.read_csv(os.path.join(input_path, input_bed), sep='\t')

df_read = df[df['ref_position'] == 1162084]
df_read.shape[0]
73

df_new = df_read[(df_read['mod_qual'] >= 0.8) | (df_read['mod_qual'] <= 0.2)]
df_new.shape[0]
68
```

**Comment by ArtRand** [comment]
Hello @PanZiwei 

Are you using the script `1.1_modkit_pileup_persite.sh` to generate the `modkit pileup` results? In this file, this command

```bash
modkit pileup -t ${THREAD} --cpg --ref ${ref_fasta} --no-filtering --mod-threshold m:0.8 ${input_bam} ${temp_bed}
```

the `--no-filtering` flag will disable the `--mod-threshold` option. This shouldn't be allowed, but there is a log line:

```
> not performing filtering
```

If I change this command to 

```
# tested with 0.1.12 and 0.1.13
modkit pileup test/subset.1162084.bam - --include-bed test/subset.1162084_include.bed --mod-threshold m:0.8 --filter-threshold 0.8
```
The resultant BED is:
```
custom_dna      1162084 1162085 m       68      +       1162084 1162085 255,0,0 68 16.18 11 57 0 0 5 0 1
```
Which I believe is the same as your `modkit extract` result:

```
custom_dna	1162084	1162085 0.161765	+	68	11 57
```
The contents of `subset.1162084_include.bed` are:

```text
custom_dna  1162084 1162085 .   100 +
```

By the way, I updated the python code above. To mimic modkit, a base modification probability of 0.8 would be called 5mC. However, due to how the base modification probabilities are discretized the base modification probability can not be exactly 0.80, 204 is 0.7988281 and 205 is 0.8027344.

**Comment by PanZiwei** [comment]
Thanks for the testing! It is a typo and I rerun the command with

**Comment by PanZiwei** [comment]
Hi,
Thanks for checking! It was a typo when I prepared the testing dataset. I rerun the `modkit pileup` but got a different result from you...
```
modkit pileup -t ${THREAD} --cpg --ref ${ref_fasta} --mod-threshold m:0.8 --mod-threshold m:0.8 ${input_bam} ${temp_bed}
awk 'BEGIN{OFS="\t"} {$5=$11; $7=$10; $8=$12; $9=$13; NF=9; print}' ${temp_bed} > ${output_bed}
```

```
## modkit pileup
custom_dna	1162084	1162085 16.44	+	38	11	27
custom_dna	1162085	1162086	9.59	 -	73	7	66
custom_dna	1162145	1162146	9.86	+	71	7	64
custom_dna   1162146   1162147	14.63	-	82	12	70
custom_dna  1162192 1162193	11.29	+	62	7	55

// vs
## modkit extract
custom_dna	1162084	1162085 0.161765	+	68	11 57
custom_dna	1162085	1162086	0.092105	 -	76	7  69	
custom_dna	1162145	1162146	0.094595	+	74	7   67
custom_dna   1162146   1162147   0.133333 -  90  12 78
custom_dna  1162192 1162193  0.101449 +  69  7  62

```

**Comment by ArtRand** [comment]
Hello @PanZiwei, 

In your command:
```
modkit pileup -t ${THREAD} --cpg --ref ${ref_fasta} --mod-threshold m:0.8 --mod-threshold m:0.8 ${input_bam} ${temp_bed}
```
you've specified `--mod-threshold` twice for `m`. In order to specify the 0.8/0.2 threshold you need `--mod-threshold m:0.8 --filter-threshold C:0.8`. If you don't specify `--filter-threshold C:0.8` (or `--filter-threshold 0.8`) `modkit` will estimate the default threshold for you, and from your data it looks like the estimate will be higher than what you want. Just a note that `--mod-threshold C:0.8` will set the threshold for mod-code `C` which is _any C mod_ [not canonical](https://samtools.github.io/hts-specs/SAMtags.pdf). 

In the command you have posted, `modkit` will estimate the filter threshold for canonical C, and use the provided threshold of >=8.0 for 5mC (`m`). When I use your command, I see the following log lines:

```
> Using filter threshold 0.9824219 for C.
> Using filter threshold 0.8 for mod code m.
```

As the log states, the filter threshold for C will be 0.98, a bit higher than you're intending. The increased threshold is the reason for the decrease in valid coverage.


In general, I'd recommend using the `--log-filepath` option to capture what's going on, especially if for any reason you're likely to loose the terminal output. 


When I use this command with the data you've provided:
```
./modkit pileup test/subset.1162084.bam - --cpg --ref test/custom_dna.fasta --filter-threshold C:0.8 --mod-threshold m:0.8 --log-filepath log.txt 2> /dev/null | awk '$2>=1162084 && $2<=1162192'
```

I get the expected output:
```
custom_dna      1162084 1162085 m       68      +       1162084 1162085 255,0,0 68 16.18 11 57 0 0 5 0 1
custom_dna      1162085 1162086 m       76      -       1162085 1162086 255,0,0 76 9.21 7 69 0 0 0 3 12
custom_dna      1162145 1162146 m       74      +       1162145 1162146 255,0,0 74 9.46 7 67 0 0 0 0 0
custom_dna      1162146 1162147 m       90      -       1162146 1162147 255,0,0 90 13.33 12 78 0 0 1 0 0
custom_dna      1162192 1162193 m       69      +       1162192 1162193 255,0,0 69 10.14 7 62 0 0 1 0 3
```

Another quick note, there might be a bug somewhere in your analysis of the `modkit pileup` results, this line
```
custom_dna	1162084	1162085 16.44	+	38	11	27
```
is incorrect (11 / 38) * 100 is 28.95% not 16.44%.

**Comment by PanZiwei** [comment]
Hi @ArtRand 
Thanks for the detailed instructions.
1. The 2 methods gave the same result when I adjusted the method 1 with `--mod-threshold m:0.8 --filter-threshold C:0.8`. The suggestion is really helpful to set up the log text to better debug the issue. 

3. It is a typo when I copy and paste the result into GitHub. The result I got when using ``--mod-threshold m:0.8` is:
```
custom_dna	1162084	1162085 28.95	+	38	11	27
```

I will close the issue now since the question is finally resolved. Thanks again for your detailed instructions and debugging during the process. Really appreciate it!
